File: LeetCode/!0022. Generate Parentheses.cpp
/*
0022. Generate Parentheses

给定n对括号，写一个函数生成所有有效的括号组合。例如，给定n=3，解集为：
[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]

思路：回溯法（深度优先遍历结果树进行枚举 + 按题目要求剪枝）
满足条件：左右括号数量相等，都为n。当左括号小于右括号数量，或者左/右括号数量大于n，则停止深度探索，return
*/

class Solution {
    public:
        void dfs(vector<string>& result, string current, int n, int left_bracket, int right_bracket) {
            // 如果是这几种情况，则不再向下深度遍历
            if (left_bracket < right_bracket || left_bracket > n || right_bracket > n)
                return;
            // 满足条件，加入解集
            if (left_bracket == right_bracket && left_bracket == n) {
                result.push_back(current);
                return;
            }
            // dfs遍历
            dfs(result, current + '(', n, left_bracket + 1, right_bracket);
            dfs(result, current + ')', n, left_bracket, right_bracket + 1);
        }

        vector<string> generateParenthesis(int n) {
            vector<string> result;
            dfs(result, "", n, 0, 0);
            return result;
        }
};

File: LeetCode/!0022. Generate Parentheses.py
'''
0022. Generate Parentheses

数字 n 代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且 有效的 括号组合。

思路：回溯法。要保证：
1）现有的左括号数量必须保持大于等于右括号数量；
2）最后的左右括号数量相等，等于n
'''

from typing import List

class Solution:
    def __init__(self) -> None:
        super().__init__()
        self.res = []
        self.cur_combination = ''

    def backtrack(self, n, left, right):
        if left == n and left == right:
            self.res.append(self.cur_combination)
            return
        # 增加左括号
        if left < n:
            self.cur_combination += '('
            self.backtrack(n, left + 1, right)
            # 回溯
            self.cur_combination = self.cur_combination[:-1]
        # 增加右括号
        if left > right:
            self.cur_combination += ')'
            self.backtrack(n, left, right + 1)
            # 回溯
            self.cur_combination = self.cur_combination[:-1]
            

    def generateParenthesis(self, n: int) -> List[str]:
        self.backtrack(n, 0, 0)
        return self.res
